### ISSUE TITLE
[Bug]: Navigation timeout of 30000 ms exceeded with Puppeteer

### ISSUE THREAD

**Issue by saldous** [puppeteer/puppeteer#8829]
### Bug description

Firebase functions 3.18.0
Puppeteer 16.2.0
Node 16

Works fine in emulator, as soon as I deploy to the cloud it gives "Navigation timeout of 30000 ms exceeded" error 90% of the time just trying to launch the browser. I've even tried increasing the memory to 8GB and makes no difference.

```
const functions = require("firebase-functions");
const puppeteer = require("puppeteer");

exports.test = functions
.runWith ({memory: "512MB"})
.https
.onRequest(async (request, response) => {

    var browser = await puppeteer.launch({
        args: [ '--no-sandbox', '--disable-setuid-sandbox',],
        headless: true,
     });

    try {
        const page = await browser.newPage();
                
        await page.goto("https://example.com");

        // SEND RESPONSE
            return response.send(JSON.stringify({
                message : "success" }));
        

    } catch (e) {
        return response.send(
            JSON.stringify({
                error : true,
                errorMessage : e.message
            })
        );
    } finally {
        await browser.close();
    } 
});
```

### Puppeteer version

16.2.0

### Node.js version

16

### npm version

8.5.0

### What operating system are you seeing the problem on?

macOS

### Relevant log output

```shell
TimeoutError: Navigation timeout of 30000 ms exceeded
    at LifecycleWatcher._LifecycleWatcher_createTimeoutPromise (/workspace/node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:164:12)
    at async Frame.goto (/workspace/node_modules/puppeteer/lib/cjs/puppeteer/common/Frame.js:169:21)
    at async Page.goto (/workspace/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1144:16)
    at async /workspace/index.js:73:9
```

**Comment by OrKoN** [comment]
Could you please provide a log that is produced when the env variable `DEBUG` is set to `puppeteer:*`? Not sure how to set an env var in Cloud functions but there must be a way?

**Comment by saldous** [comment]
I'll look into how to do this, this is what I see in the Firebase logs:
<img width="1200" alt="image" src="https://user-images.githubusercontent.com/3334306/186410716-b919d5ad-02a2-4c47-b27c-74fc5b54cff3.png">

**Comment by saldous** [comment]
//This works more often than not, but doesn't catch timeout and causes crashes... I would rather have a crash and more working sessions...
        `await page.goto(`https://example.com`, { waitUntil: 'load' });`

//THIS SHOULD WORK! BUT 80% OF THE TIME IT IS CATCHING A TIMEOUT. IS IT FIREBASE ISSUE OR PUPPETEER?
```
 try {
            await page.goto(`https://example.com`, { waitUntil: 'load' });
            console.log("SUCCESS!! WOOHOO");
        } catch (e) {
            console.log("Try/Catch error: ", e);
            await browser.close();
        }
```

**Comment by OrKoN** [comment]
@saldous have you tried to increase the timeout? If it happens only on Firebase it's likely that the runtime environment is much slower and/or throttling/blocking network or smth else going wrong in that environment.

**Comment by qrerve01** [comment]
I am having the exact same experience.   When it works, the response comes back in about 4 seconds (a reasonable amount of time for this particular page to load).  when it does not come back, it fails after the 30 seconds.  it is very random and I am currently always fetching the same page which is very basic.   it either comes back in 4 seconds or never comes back, nothing in between.
Mine is deployed on Dreamhost on a VPS
the only difference in my code is that I am waiting on domcontentloaded vs just load.

here is the error:  (masking my file system with "......")

.......node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:164
    return new Errors_js_1.TimeoutError(errorMessage);
           ^

TimeoutError: Navigation timeout of 30000 ms exceeded
    at LifecycleWatcher._LifecycleWatcher_createTimeoutPromise (........node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:164:12)

Chromium version:  HeadlessChrome/105.0.5173.0

+-- puppeteer@16.2.0

I also tried to downgrade to 16.1.1 and had the same experience, which seems really odd.

**Comment by bluepeter** [comment]
Also having this issue.

**Comment by saldous** [comment]
> @saldous have you tried to increase the timeout? If it happens only on Firebase it's likely that the runtime environment is much slower and/or throttling/blocking network or smth else going wrong in that environment.

Increasing timeout makes no difference. Just fails after 1min rather than 30seconds

**Comment by qrerve01** [comment]
> > @saldous have you tried to increase the timeout? If it happens only on Firebase it's likely that the runtime environment is much slower and/or throttling/blocking network or smth else going wrong in that environment.
> 
> Increasing timeout makes no difference. Just fails after 1min rather than 30seconds

Concur, this is definitely not a timeout issue as per the description - it either works in about 4 seconds or never works, regardless of the timeout.

**Comment by saldous** [comment]
Simple function, works on my machine, timeout 99% of the time when deployed.
```
const PUPPETEER_OPTIONS = {
    headless: true,
    args: [
        '--disable-gpu',
        '--disable-setuid-sandbox',
        '--no-sandbox',
        '--no-first-run',
        '--no-zygote',
    ],
    sloMo: 50
};

const openConnection = async () => {
    const browser = await puppeteer.launch(PUPPETEER_OPTIONS);
    const page = await browser.newPage();
    await page.setUserAgent(
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36'
    );
    await page.setViewport({ width: 1000, height: 900 });
    return { browser, page };
};

const closeConnection = async (page, browser) => {
    page && (await page.close());
    browser && (await browser.close());
};


exports.test = functions
    .runWith({ memory: '1GB' })
    .https.onRequest(async (request, response) => {
            let { browser, page } = await openConnection();

            try {
                await page.goto(`https://www.example.com`, { waitUntil: ['domcontentloaded', 'networkidle2'], });

                console.log("HERE: Successfully loaded page")

                response.send(JSON.stringify({
                    error: false,
                    errorMessage: "Ok"
                }));

            } catch (err) {
                console.log(`Error: ${err.message}`);
                    response.send(JSON.stringify({
                        error: true,
                        errorMessage: err.message
                    }));
            } finally {
                await closeConnection(page, browser);
            }
    });
```

**Comment by qrerve01** [comment]
> @OrKoN I don't know how to use env var DEBUG=puppeteer.* within Firebase Function. Any guidance appreciated. Getting quite frustrated on trying to solve this issue. Thank you.

I dont know if I did it right, but here is what I did on the DEBUG side - I should mention that I get DEBUG messages, but not from PUPPETEER.

I put this line in a file called ".env"
I put the file in a directory called environment (which in my case is at the same level as js and PHP directories
in my case my .js file is called server_ws.js but you should use your filename.

**DEBUG="puppeteer:*"  DEBUG_COLORS=true node ./js/server_ws.js 2>&1 | grep -v '"Network'**

I put this line in my .js file that I am running with node

**require('dotenv').config({ path: './environment/.env' })**

I am also using express in this node js program and that is the only think I see:   (I am getting messages - but again, no Puppeteer messages at all) 
here is sample output:

 express:application set "x-powered-by" to true +0ms
  express:application set "etag" to 'weak' +1ms
  express:application set "etag fn" to [Function: generateETag] +1ms
  express:application set "env" to 'development' +1ms
  express:application set "query parser" to 'extended' +0ms
  express:application set "query parser fn" to [Function: parseExtendedQueryString] +0ms
  express:application set "subdomain offset" to 2 +0ms
  express:application set "trust proxy" to false +1ms
  express:application set "trust proxy fn" to [Function: trustNone] +0ms
  express:application booting in development mode +0ms
  express:application set "view" to [Function: View] +0ms
  express:application set "views" to '/home/jobs_prod/alljoblistings.com/public/views' +1ms
  express:application set "jsonp callback name" to 'callback' +0ms
  express:router use '/' query +65ms
  express:router:layer new '/' +0ms
  express:router use '/' expressInit +1ms
  express:router:layer new '/' +0ms
  express:router:route new '/.websocket' +0ms
  express:router:layer new '/.websocket' +1ms
  express:router:route get '/.websocket' +0ms
  express:router:layer new '/' +0ms



> ### Bug description
> 
> Firebase functions 3.18.0 Puppeteer 16.2.0 Node 16
> 
> Works fine in emulator, as soon as I deploy to the cloud it gives "Navigation timeout of 30000 ms exceeded" error 90% of the time just trying to launch the browser. I've even tried increasing the memory to 8GB and makes no difference.
> 
> ```
> const functions = require("firebase-functions");
> const puppeteer = require("puppeteer");
> 
> exports.test = functions
> .runWith ({memory: "512MB"})
> .https
> .onRequest(async (request, response) => {
> 
>     var browser = await puppeteer.launch({
>         args: [ '--no-sandbox', '--disable-setuid-sandbox',],
>         headless: true,
>      });
> 
>     try {
>         const page = await browser.newPage();
>                 
>         await page.goto("https://example.com");
> 
>         // SEND RESPONSE
>             return response.send(JSON.stringify({
>                 message : "success" }));
>         
> 
>     } catch (e) {
>         return response.send(
>             JSON.stringify({
>                 error : true,
>                 errorMessage : e.message
>             })
>         );
>     } finally {
>         await browser.close();
>     } 
> });
> ```
> 
> ### Puppeteer version
> 
> 16.2.0
> ### Node.js version
> 
> 16
> ### npm version
> 
> 8.5.0
> ### What operating system are you seeing the problem on?
> 
> macOS
> ### Relevant log output
> 
> ```shell
> TimeoutError: Navigation timeout of 30000 ms exceeded
>     at LifecycleWatcher._LifecycleWatcher_createTimeoutPromise (/workspace/node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:164:12)
>     at async Frame.goto (/workspace/node_modules/puppeteer/lib/cjs/puppeteer/common/Frame.js:169:21)
>     at async Page.goto (/workspace/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:1144:16)
>     at async /workspace/index.js:73:9
> ```

Another data point,  I am calling my node js program via websocket.  when I get a timeout from puppeteer, it sends a message back via websocket that says it failed.  At that point, I immediately resend the request with the same data.  sometimes that works, sometimes it does not.   At no point have I had to send more than 3 messages before it works.   the timeout just happens when it feels like it.   I am literally calling the exact same page, parameters etc until it works.   In my case, it works about 80% of the time but when it fails, this is how I have worked around the issue for now.   Definitely a HACK, but it allowed me to keep developing with the hope that a future release will fix this.

**Comment by OrKoN** [comment]
@qrerve01 that's weird, do you get the debug logs if you run your program locally? I think if you load env from inside NodeJS you need make sure it happens before you require Puppeteer.
@saldous Unfortunately, I have never used Firebase functions and cannot provide further help.

**Comment by saldous** [comment]
Update: I moved my Firebase Function from 'us-central1' region to 'us-east4' region.. and now it works, same code. So the issue must have been with Firebase's network in us-central1.

**Comment by Hotell** [comment]
We are running into same issues for quite some time now https://uifabric.visualstudio.com/fabricpublic/_build/results?buildId=283275&view=logs&j=258ec178-2d8b-5611-7b9b-60c5c95dae55&t=254ee48e-1503-524f-f75e-e2c69d0380f5

search for `@fluentui/ssr-tests-v9 test` in the logs.

We have retry mechanism in place as well which provides no benefit, once it fails first time after 4 more attempts fails anyways with the same error https://github.com/microsoft/fluentui/pull/26375

**Comment by ryggdunk** [comment]
I have the exact same issue as original poster. Works perfect on local machine but as soon as I run the script on my cloud server it fails after 30 seconds. I managed to make it work once or twice but that's about it. Increasing time before timeout doesn't help. Just increases time before I get the timeout.

**Comment by voxvici** [comment]
Same issue here

**Comment by kazeens** [comment]
> I have the exact same issue as original poster. Works perfect on local machine but as soon as I run the script on my cloud server it fails after 30 seconds. I managed to make it work once or twice but that's about it. Increasing time before timeout doesn't help. Just increases time before I get the timeout.

Have you found a solution?

**Comment by kazeens** [comment]
Having this random timeout issue on my VPS. (local is fine)
What might be the problem?
cc @OrKoN

**Comment by fuelkoy** [comment]
Check this answer from another issue: https://github.com/puppeteer/puppeteer/issues/4796#issuecomment-535612842.
Whatever place you host your function make sure your function responds only after Puppeteer is done.

**Comment by younesben197** [comment]
Disable timeout
await page.goto(url, { waitUntil: 'load', timeout: 0 });

**Comment by razvanandreib** [comment]
I've had a similar problem loading a script on my local machine and added added the timeout to 0 and wait until domcontentloaded:

    await page.goto(url, { timeout: 0, waitUntil: "domcontentloaded" })

that appears to work for me.

**Comment by e-boudhina** [comment]
I am having the same issue:
 await page.goto(AutoScout24URL, { waitUntil: 'networkidle2' }); did not work, I also tried to use "await page.setDefaultNavigationTimeout(0);" still nothing...

**Comment by kyungseopk1m** [comment]
This issue was very troublesome for me. 
It would proceed normally, but occasionally the error would occur again. 
So, I added the following code before the page.goto and was able to resolve the problem:
`page.setDefaultNavigationTimeout(0);`

**Comment by gustavovni08** [comment]
any solution??

**Comment by hafizab63861462** [comment]
async function initializeBrowser() {
  try {
    const browser = await puppeteer.launch({
      headless: true,
      defaultViewport: null,
      args: [
        "--no-sandbox",
        "--disable-setuid-sandbox",
        "--disable-dev-shm-usage",
        "--disable-accelerated-2d-canvas",
        "--no-first-run",
        "--no-zygote",
        "--disable-gpu",
        "--disable-extensions",
        "--disable-features=site-per-process",
        "--ignore-certificate-errors",
        "--disable-blink-features=AutomationControlled",
        "--disable-infobars",
        "--window-position=0,0",
        "--ignore-certifcate-errors",
        "--ignore-certifcate-errors-spki-list",
      ],
    });

    const page = await browser.newPage();

    const rootDir = path.join(__dirname, "..", "..");
    const videoDir = path.join(rootDir, "recordings");

    // Create directory if it doesn't exist
    if (!fs.existsSync(videoDir)) {
      fs.mkdirSync(videoDir, { recursive: true });
    }

    const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
    const jobId = Math.random().toString(36).substring(2, 8);
    const fileName = `recording-${jobId}-${timestamp}.mp4`;
    const savePath = path.join(videoDir, fileName);

    const recorder = new PuppeteerScreenRecorder(page);
    await recorder.start(savePath);

    return {
      browser,
      page,
      recorder,
    };
  } catch (error) {
    console.error("Error initializing browser:", error);
    throw error;
  }
}

async function performLogin(page, accountLink, email, password) {
  try {
    console.log("Navigating to login page...");
    await page.goto(accountLink, {
      waitUntil: ["load", "domcontentloaded", "networkidle2"],
      timeout: 0
    });

    console.log("Filling login form...");
    await page.waitForSelector("#emailOrUsernameInput");
    await page.type("#emailOrUsernameInput", email);

    await page.waitForSelector("#passwordInput");
    await page.type("#passwordInput", password);

    await page.waitForSelector('button[type="submit"]', { visible: true });
    await page.click('button[type="submit"]');

    await page.waitForNavigation({ waitUntil: "networkidle0" });
  } catch (error) {
    console.error("Error during login:", error);
    throw new Error(error);
  }
}


found any solution i also got same issue in render deployment and it is working fine in my local machine 

see the screen short

<img width="1044" alt="Image" src="https://github.com/user-attachments/assets/9d9ac4fd-a8d6-4359-af80-bad17250af23" />
